### PR TITLE
Added check in virt-controller to update podNetwork IPs on VMI status interface

### PR DESCRIPTION
**Pull request by vatsalparekh**
Signed-off-by: Vatsal Parekh <vparekh@redhat.com>

**What this PR does / why we need it**:
Added check in virt-controller to update podNetwork IPs on VMI status interface

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
**Fixes** # https://bugzilla.redhat.com/show_bug.cgi?id=1686208

**Special notes for your reviewer**:

**Release note**:

```release-note
Added check in virt-controller to update podNetwork interface IPs
```

### CODE REVIEW

File: pkg/virt-controller/watch/vmi.go [hunk 1]
```diff
@@ -380,6 +380,28 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
...
+				break
+			}
+		}
+		newInterfaces, err := json.Marshal(vmiCopy.Status.Interfaces)
+		if err != nil {
+			return err
+		}
+		log.Log.V(3).Object(vmi).Infof("Patching VMI interfaces")
+		patch := fmt.Sprintf(`{ "op": "replace", "path": "/status/interfaces", "value": %s }`, string(newInterfaces))
+		_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, []byte(fmt.Sprintf("[ %s ]", patch)))
```
Comment (rmohr) [inline pkg/virt-controller/watch/vmi.go]: This would always do a patch, no matter if something changed. You will also have to check if you actually changed something.

File: pkg/virt-controller/watch/vmi.go [hunk 2]
```diff
@@ -380,6 +380,28 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 			conditionManager.RemoveCondition(vmiCopy, virtv1.VirtualMachineInstanceConditionType(k8sv1.PodReady))
 		}
 
+		// IF the VMI has pod network, Check and patch if IP on VMI is not same as the one on Pod IP
+		for _, network := range vmiCopy.Spec.Networks {
+			if network.NetworkSource.Pod != nil {
+				for i, iface := range vmiCopy.Status.Interfaces {
+					if iface.Name == network.Name {
+						vmiCopy.Status.Interfaces[i].IP = pod.Status.PodIP
```
Comment (rmohr) [inline pkg/virt-controller/watch/vmi.go]: This  looks good, but we also have an IPs field. You will also have to search in the list there if it is present.
Further I think that virt-handler will always override the IP with the link-local address of the VM which is reported by the qemu guest agent. We will also have to update the status update in virt-handler so that it respects the pod IP if `masquerade` is set and only addes the link-local address to `IPs`.

Comment (phoracek) [reply pkg/virt-controller/watch/vmi.go]: The IPs field is handled on the line 390, is not it? Maybe it has been added since the comment.

We first report the pod interface with the pod IP here: https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-controller/watch/vmi.go#L339, no guest agent at this point.

The guest agent reporting is handled here https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L275. And I believe it overwrites the orignal Pod interface report here https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L315 in case the guest agent is running. It smells though, I'd expect we have that tested, it should still report the public pod IP, not the link local.

Comment (phoracek) [reply pkg/virt-controller/watch/vmi.go]: @vatsalparekh do you have a KubeVirt cluster running? Could you check what IP is reported on a VMI with masquerade when guest agent is installed?

File: pkg/virt-controller/watch/vmi.go [hunk 3]
```diff
@@ -380,6 +380,40 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 			conditionManager.RemoveCondition(vmiCopy, virtv1.VirtualMachineInstanceConditionType(k8sv1.PodReady))
 		}
 
+		var patchNeeded bool
```
Comment (fabiand) [inline pkg/virt-controller/watch/vmi.go]: Maybe a more descriptive variable name ..

File: pkg/virt-controller/watch/vmi.go (hunk 2 above)
Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go]: @petrkotas 
```
➜  kubevirt git:(migration-bindings-fix) ✗ kg vmi -o wide
NAME         AGE   PHASE     IP            NODENAME   LIVE-MIGRATABLE
vmi-fedora   22m   Running   10.0.2.2/24   node01     True
```
```
    interfaces:
    - interfaceName: eth0
      ipAddress: 10.0.2.2/24
      ipAddresses:
      - 10.0.2.2/24
      - fe80::ff:fe72:7c3c/64
      mac: 02:00:00:72:7c:3c
      name: default

```

Comment (phoracek) [reply pkg/virt-controller/watch/vmi.go]: Thanks. It is an unrelated bug. Let's tackle it in a separate PR.

File: pkg/virt-controller/watch/vmi.go [hunk 4]
```diff
@@ -380,6 +380,43 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
...
 
+		var networkPatchNeeded bool
+		// IF the VMI has a pod network, Check and patch if IP on VMI is not same as the one on Pod IP
+		for _, network := range vmiCopy.Spec.Networks {
+			if network.NetworkSource.Pod == nil {
+				continue
+			}
+			for i, iface := range vmiCopy.Status.Interfaces {
+				if iface.Name == network.Name {
+					vmiCopy.Status.Interfaces[i].IP = pod.Status.PodIP
```
Comment (danielBelenky) [inline pkg/virt-controller/watch/vmi.go]: Wouldn't this patch every time? Shouldn't you patch only if the ip on the interface is differs from the one on the pod?

File: pkg/virt-controller/watch/vmi.go [hunk 5]
```diff
@@ -380,6 +380,43 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
...
+					networkPatchNeeded = true
+
+					if iface.IPs == nil {
```
Comment (danielBelenky) [inline pkg/virt-controller/watch/vmi.go]: This check is redundant... you can iterate over `nil` fields (it just won't iterate, but it will not panic)

File: pkg/virt-controller/watch/vmi.go [hunk 6]
```diff
@@ -380,6 +380,43 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
...
+						continue
+					}
+					for j, ip := range iface.IPs {
+						if ip == iface.IP {
+							vmiCopy.Status.Interfaces[i].IPs[j] = pod.Status.PodIP
```
Comment (danielBelenky) [inline pkg/virt-controller/watch/vmi.go]: Is this correct? For example: if user sets IP manually, after migration we change all the IPs on a given interface to the Pod's IP?

File: tests/migration_test.go [hunk 7]
```diff
@@ -1571,9 +1572,32 @@ var _ = Describe("[rfe_id:393][crit:high[vendor:cnv-qe@redhat.com][level:system]
...
+						vmiPod := tests.GetRunningPodByVirtualMachineInstance(newvmi, tests.NamespaceTestDefault)
+
+						for _, network := range newvmi.Spec.Networks {
+							if network.NetworkSource.Pod != nil {
+								for _, iface := range newvmi.Status.Interfaces {
+									if iface.Name == network.Name {
+										if iface.IP == vmiPod.Status.PodIP {
+											return nil
+										} else {
+											return goerrors.New(fmt.Sprintf("VMI IP:%s not equal to Pod IP:%s", iface.IP, vmiPod.Status.PodIP))
```
Comment (danielBelenky) [inline tests/migration_test.go]: You could use `fmt.Errorf`here instead of explicitly creating a new error... just a matter of readability IMO.

File: pkg/virt-controller/watch/vmi.go (hunk 6 above)
Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go]: I'm not sure why/how user would set Pod IP manually? is it possible to change that?
We're just changing the IP on Interface with Pod network to the IP of the new migrated virt-launcher pod.
@rmohr @phoracek your views?

File: tests/migration_test.go (hunk 7 above)
Comment (vatsalparekh) [reply tests/migration_test.go]: Changed.

File: pkg/virt-controller/watch/vmi.go (hunk 5 above)
Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go]: Removed

File: pkg/virt-controller/watch/vmi.go (hunk 4 above)
Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go]: Ah yes, good catch, changed now.

File: pkg/virt-controller/watch/vmi.go (hunk 6 above)
Comment (rmohr) [reply pkg/virt-controller/watch/vmi.go]: the podIP can't be set by users. Right now in theory users can mess up our VMI status because we don't protect the status struct yet. We have to enable at some point the status subresource support. But anyway, our controllers owns status completely and can set or override there everything.

File: tests/migration_test.go [hunk 8]
```diff
@@ -1571,9 +1571,32 @@ var _ = Describe("[rfe_id:393][crit:high[vendor:cnv-qe@redhat.com][level:system]
...
+
+				By("Checking that all migrated VMIs have the new pod IP address")
+				for _, vmi := range vmis {
+					Eventually(func() error {
+						newvmi, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.Name, &metav1.GetOptions{})
+						Expect(err).ToNot(HaveOccurred())
+						vmiPod := tests.GetRunningPodByVirtualMachineInstance(newvmi, tests.NamespaceTestDefault)
+
+						for _, network := range newvmi.Spec.Networks {
+							if network.NetworkSource.Pod != nil {
```
Comment (oshoval) [inline tests/migration_test.go]: sorry in advance if i mistaken,

the code has a break 
"// Currently will only traverse to the first Pod network found"
the test doesnt in case there is more than one pod network found
is it intentional or need to add a break on the first one found ?

File: pkg/virt-controller/watch/vmi.go [hunk 9]
```diff
@@ -380,6 +380,42 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
...
+
+						for j, ip := range iface.IPs {
+							if ip == iface.IP {
+								vmiCopy.Status.Interfaces[i].IPs[j] = pod.Status.PodIP
+							}
+						}
+					}
+				}
+			}
+			// Currently will only traverse to the first Pod network found
```
Comment (oshoval) [inline pkg/virt-controller/watch/vmi.go]: we are ok with it ? might be an errata in the future ?

**Comment by fabiand** [comment]
ping @vatsalparekh ?

File: pkg/virt-controller/watch/vmi.go (hunk 9 above)
Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go]: my understanding with multiple pod networks is limited here, @oshoval can you describe more here?

Comment (oshoval) [reply pkg/virt-controller/watch/vmi.go]: tbh mine as well, 
thought just generally what happens if there are more than one pod network, 
do we need to support this feature across all of them ?

File: pkg/virt-controller/watch/vmi.go [hunk 10]
```diff
@@ -380,6 +380,42 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
...
+				if iface.Name == network.Name {
+					if vmiCopy.Status.Interfaces[i].IP != pod.Status.PodIP {
+						vmiCopy.Status.Interfaces[i].IP = pod.Status.PodIP
+						networkPatchNeeded = true
+
+						for j, ip := range iface.IPs {
+							if ip == iface.IP {
+								vmiCopy.Status.Interfaces[i].IPs[j] = pod.Status.PodIP
+							}
+						}
```
Comment (vladikr) [inline pkg/virt-controller/watch/vmi.go]: Aren't these addresses being reported by the agent?
We will overwrite the actual guest addresses if we will do it here.
Perhaps we should only do it if the agent is not configured?
@phoracek what do you think?

File: pkg/virt-controller/watch/vmi.go [hunk 11]
```diff
@@ -380,6 +380,42 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 			conditionManager.RemoveCondition(vmiCopy, virtv1.VirtualMachineInstanceConditionType(k8sv1.PodReady))
 		}
 
+		var networkPatchNeeded bool
+		// IF the VMI has a pod network, Check and patch if IP on VMI is not same as the one on Pod IP
+		for _, network := range vmiCopy.Spec.Networks {
```
Comment (vladikr) [inline pkg/virt-controller/watch/vmi.go]: I know we're only looking for a single pod network address, but 3 `for`s... I would at least break the network lookup to a variable. (or a map if you expect more than one pod network)

Comment (phoracek) [reply pkg/virt-controller/watch/vmi.go]: There is only a single Pod network. I think we are safe counting on that.

File: pkg/virt-controller/watch/vmi.go (hunk 10 above)
Comment (phoracek) [reply pkg/virt-controller/watch/vmi.go]: To be honest, I don't fully get why we need to handle this in watch/vmi.go and not in handler as in https://github.com/kubevirt/kubevirt/pull/3063/files.

File: pkg/virt-controller/watch/vmi.go (hunk 11 above)
Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go]: @vladikr Yes it's not a big look but good suggestion to improve it, moved the network lookup to a map

File: pkg/virt-controller/watch/vmi.go (hunk 10 above)
Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go]: @phoracek @vladikr I can be wrong here, but I guess `virt-controller` already has the Pod details, so it might be straightforward to update this from here

Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go]: @vladikr about guest-agent, this PR is doing the change to only update agent address only if it's not masquerade https://github.com/kubevirt/kubevirt/pull/3063/files
Are you suggesting any other change here?

**Comment by fabiand** [comment]
@vatsalparekh ping?

**Comment by vatsalparekh** [comment]
@fabiand this cannot be merged without https://github.com/kubevirt/kubevirt/pull/3063/
With only this patch, we will have a race condition on the VMI status where both virt-handler and virt-controller will try to update it when guest-agent is present

**Comment by fabiand** [comment]
Ah, thanks for the explanation!

File: tests/migration_test.go [hunk 12]
```diff
@@ -1575,6 +1575,21 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 					targetNode.Name,
 					targetNode.Name,
 				))
+
+				By("Checking that all migrated VMIs have the new pod IP address on VMI status")
+				for _, vmi := range vmis {
+					Eventually(func() error {
+						newvmi, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmi.Name, &metav1.GetOptions{})
+						Expect(err).ToNot(HaveOccurred())
```
Comment (phoracek) [inline tests/migration_test.go:1583]: Please add an expectation string at the end of the expect to make test debugging easier.

File: pkg/virt-controller/watch/vmi.go [hunk 13]
```diff
@@ -380,6 +380,40 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 			conditionManager.RemoveCondition(vmiCopy, virtv1.VirtualMachineInstanceConditionType(k8sv1.PodReady))
 		}
 
+		var networkPatchNeeded bool
+		existingNetworks := map[string]virtv1.Network{}
```
Comment (phoracek) [inline pkg/virt-controller/watch/vmi.go:384]: What is an `existingNetwork`? It sounds a little too generic and does not help me understand what this keeps. Maybe `vmiNetworksByName` would be better?

File: pkg/virt-controller/watch/vmi.go [hunk 14]
```diff
@@ -380,6 +380,40 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
...
+			}
+		}
+		for i, iface := range vmiCopy.Status.Interfaces {
+			if existingNetworks[iface.Name].NetworkSource.Pod != nil && vmiCopy.Status.Interfaces[i].IP != pod.Status.PodIP {
+				vmiCopy.Status.Interfaces[i].IP = pod.Status.PodIP
+				networkPatchNeeded = true
+
+				for j, ip := range iface.IPs {
+					if ip == iface.IP {
+						vmiCopy.Status.Interfaces[i].IPs[j] = pod.Status.PodIP
```
Comment (phoracek) [inline pkg/virt-controller/watch/vmi.go:397]: Since we report only the PodIP in IPs with masquerade, we should not really need to iterate all the IPs. It should be sufficient to only se the IPs to [IP]. Am I missing some scenario when this is not the case?

File: pkg/virt-controller/watch/vmi.go [hunk 15]
```diff
@@ -380,6 +380,40 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
...
+					}
+				}
+				// Currently will only traverse to the first Pod network found
+				break
+			}
+		}
+		if networkPatchNeeded {
```
Comment (phoracek) [inline pkg/virt-controller/watch/vmi.go:404]: On line 417 and 431 we have some comments regarding the patching. Would it make sense to have them here as well?

File: tests/migration_test.go (hunk 12 above)
Comment (vatsalparekh) [reply tests/migration_test.go:1583]: Added

File: pkg/virt-controller/watch/vmi.go (hunk 15 above)
Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go:404]: Actually, this can be changed to use the new `patchOps` and Conditions checks on latest code.
Changed it now.

File: pkg/virt-controller/watch/vmi.go [hunk 16]
```diff
@@ -390,6 +390,41 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
...
+				vmiNetworksByName[network.Name] = network
+			}
+		}
+		for i, iface := range vmiCopy.Status.Interfaces {
+			if vmiNetworksByName[iface.Name].NetworkSource.Pod != nil && vmiCopy.Status.Interfaces[i].IP != pod.Status.PodIP {
+				vmiCopy.Status.Interfaces[i].IP = pod.Status.PodIP
+
+				// For forwarding binding mechanism like masquerade,
+				// it's sufficient to expose only PodIP to list of IPs
+				vmiCopy.Status.Interfaces[i].IPs = []string{pod.Status.PodIP}
```
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:405]: don't we still have two components (virt-handler and virt-controller) simultaneously attempting to set this same value? Doesn't this conflict?

Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go:405]: @davidvossel I don't think so
1. I see that virt-handler only reuses the interfaces if they're present in vmi status https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L451
2. If Migrating VMI has Guest-agent, that is already handled and won't be updating from virt-handler  https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L470
3. Migrations are allowed with only masquerede so I don't see any other conflicts happening in theory https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/vm.go#L1619

Still, do you see any other places that can cause conflict?

Comment (davidvossel) [reply pkg/virt-controller/watch/vmi.go:405]: > Still, do you see any other places that can cause conflict?

it's difficult to tell, and the risk is high here. 

The `vmi.Status.Interfaces` array is written to by virt-handler when the VMI is in the Running phase, which means that virt-handler effectively owns those fields.

If there's ever an instance where somehow virt-handler and virt-controller are not in alignment on what value should be set on this field, we'll get into a reconcile loop storm where virt-handler and virt-controller continually attempt to override each others values.

Basically, this logic is complex enough that even if it's right today, and you've gotten every if statement separated between what values virt-handler and virt-controller sets, it's still fragile logic. We run the risk of breaking this or making a subtle change to the logic in the future which could impact all of KubeVirt. 

Is it possible to get the same results by setting the interface IP from virt-handler instead of virt-controller?

Comment (vatsalparekh) [reply pkg/virt-controller/watch/vmi.go:405]: > > Still, do you see any other places that can cause conflict?
> 
> it's difficult to tell, and the risk is high here.
> 
> The `vmi.Status.Interfaces` array is written to by virt-handler when the VMI is in the Running phase, which means that virt-handler effectively owns those fields.
> 
> If there's ever an instance where somehow virt-handler and virt-controller are not in alignment on what value should be set on this field, we'll get into a reconcile loop storm where virt-handler and virt-controller continually attempt to override each others values.
> 
> Basically, this logic is complex enough that even if it's right today, and you've gotten every if statement separated between what values virt-handler and virt-controller sets, it's still fragile logic. We run the risk of breaking this or making a subtle change to the logic in the future which could impact all of KubeVirt.
> 
> Is it possible to get the same results by setting the interface IP from virt-handler instead of virt-controller?
While I'm not sure if we can get Pod info in virt-handler, I don't think if it's a good practice to fetch Pod details from every virt-handler. For a single VM Migrated in the cluster, if the whole cluster's virt-handler hits api/cache for pod details, doesn't seem too promising to me.
Unless we have some way to tell that specific virt-handler to update VMI where it's getting transferred to.

your thoughts? @vladikr @rmohr @fabiand

**Comment by stu-gott** [comment]
Has this effort stalled? There hasn't been any activity in a considerable amount of time.

**Comment by vatsalparekh** [comment]
@stu-gott It was decided that virt-handler seems a better place for this fix, this can be closed as the #3642 should be the fix addressing this issue now.
(Previously #3488 was addressing this issue instead of this, but it had to be reverted after CI merged it without all tests passing)
